itou/users/factories.py

```diff
@@ -215,6 +221,14 @@ class Meta:
     # JobSeeker are created with a Pôle emploi ID
     pole_emploi_since = AllocationDuration.MORE_THAN_24_MONTHS
 
+    @classmethod
+    def _create(cls, model_class, **kwargs):
+        # User creation creates a profile
+        # So we can't use manager.create and need to update it instead
+        user = kwargs.pop("user")
+        manager = cls._get_manager(model_class)
+        return manager.update_or_create(user=user, defaults=kwargs)[0]
```

Comment:
Je ne sais pas si autoriser la modification d'objets déjà existants via une _factory_ est un chemin qu'on souhaite emprunter :thinking:.
Un peu l'impression qu'on va à contre sens du fonctionnement (de la philosophie ?) général, et donc qu'on se retrouve à faire des choses auxquelles personne ne s'attend, après c'est du _one-to-one_ donc il n'y a peut-être pas de cas d'utilisation problématique :shrug:.

Reply:
Effectivement, vu comme ça :grimacing: .
Alors je vois 2 solutions:
- forcer l'utilisation d'une `JobSeekerFactory(without_profile=True)` via:
```
-    user = factory.SubFactory(JobSeekerWithAddressFactory)
+    user = factory.SubFactory(JobSeekerWithAddressFactory, without_profile=True)
```
- utiliser la version suggérée par factory-boy https://factoryboy.readthedocs.io/en/stable/recipes.html#example-django-s-profile mais utilisant des signaux :cold_face: 

Clairement je préfère la première solution :grin:

Reply:
Sans vouloir mettre les pieds dans un plat déjà conséquent, je me disais immédiatement à la première proposition que ce `without_profile` était problématique.
Si on crée un utilisateur de type `job_seeker` alors il devrait avoir immédiatement un `JobSeekerProfile` sauf cas très particulier (auquel cas le test pourrait le supprimer a posteriori explicitement)

Ce qui reviendrait à corriger ou supprimer les tests qui cherchent à ne pas avoir de profil. Non ?

Reply:
> Si on crée un utilisateur de type job_seeker alors il devrait avoir immédiatement un JobSeekerProfile sauf cas très particulier (auquel cas le test pourrait le supprimer a posteriori explicitement)

C'est le cas dans cette PR: si tu n'as pas rajouté un `without_profile` dans l'appel de ta factory tu auras un `JobSeekerProfile`.
Et c'est encore le cas, avec ma dernière modif où la `JobSeekerProfileFactory` commence par créer un `User` avec `without_profile=True` avant de se rajouter (et l'`User` a donc bien un `JobSeekerProfile` à la fin :sweat_smile: )

> Ce qui reviendrait à corriger ou supprimer les tests qui cherchent à ne pas avoir de profil. Non ?

Pour moi, cela sera à faire mais seulement une fois la _Grande Migration_ effectuée.
Car pour le moment, on a *beaucoup* de `job_seeker` sans profil donc ça me parait important de continuer à vérifier que le code les gère bien (via des tests).

Reply:
Yes, en parcourant le code je vois ce que tu veux dire et en effet cela revient au même à peu de choses près. Soit la mention `without_profile` supprime le profil et laisse du coup le "slot" vacant, ce qui permet au test de rester écrit à peu près comme auparavant, soit (ce que je pensais faire) on ne met pas de mention et on modifie le test pour agit sur `user.jobseeker_profile` après l'execution de la Factory.

Tomato tomato, mais dans ton cas ça fait reprendre moins de code.